fusequery/query/src/servers/server.rs

```diff
@@ -65,8 +61,13 @@ impl ShutdownHandle {
 
             log::info!("Received termination signal.");
             log::info!("You can press Ctrl + C again to force shutdown.");
-            let shutdown_services = self.shutdown(Some(receiver));
-            shutdown_services.await;
+            if let Ok(false) =
+                self.shutdown
+                    .compare_exchange(false, true, Ordering::SeqCst, Ordering::Acquire)
+            {
+                let shutdown_services = self.shutdown(Some(receiver));
+                shutdown_services.await;
+            }
```

Comment:
@SGZW 
The condition check on `self.shutdown:Arc<AtomicBool>` should be the first step of `self.shutdown()`. 
Since `self` is a `&mut self`, no Arc or Atomic is required.

Reply:
atomic is to avoid race, but Arc is not required

Reply:
Both usages of `self.shutdown` are in a scope that holds a `&mut self`. 
There can not be two threads both holding `&mut self` at any time.

The only way to let another thread have a `&mut self`(at a different time) is to send the `self` through some channel.

So, IMHO there should not be race condition. (hoping I did not miss anything 🤔 )

Reply:
in this case,  `&mut self ` move to closure,  is copy, in my sense

Reply:
That's why I stated that to move the check on `self.shutdown` into `self.shutdown()`, which would simplify this issue...

Sorry I may not have expressed it very clear...

> The condition check on `self.shutdown:Arc<AtomicBool>` should be the first step of `self.shutdown()`. 
Since `self` is a `&mut self`, no Arc or Atomic is required.

Reply:
i think check outside can reduce overhead about moving variables into closure, and panic will call drop will happen in different thread?

Reply:
Hm... what about this way: 
https://github.com/datafuselabs/datafuse/blob/dd03f129e77405c1540d2f231af71d1f9e409281/common/stoppable/src/stop_handle.rs#L34-L56

I didn't get it. It should drop-in-different-thread or not?

Reply:
i get this,  all stopping flag check is in one thread by "wait_to_terminate" function move